File: Aqua Guard/frontend/app/src/main/java/com/aquaguard/presentation/devices/DevicesViewModel.java
package com.aquaguard.presentation.devices;

import android.os.Handler;
import android.os.Looper;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.aquaguard.domain.model.Device;
import com.aquaguard.domain.repository.DeviceRepository;
import com.aquaguard.domain.repository.RepositoryCallback;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class DevicesViewModel extends ViewModel {
    private final DeviceRepository deviceRepository;

    private final LiveData<List<Device>> devices;
    private final MutableLiveData<DeviceActionState> deviceActionState = new MutableLiveData<>(DeviceActionState.idle());
    private final MutableLiveData<ProvisioningState> provisioningState = new MutableLiveData<>(ProvisioningState.idle());
    private final MutableLiveData<Map<String, OtaState>> otaState = new MutableLiveData<>(new HashMap<>());

    private final Handler handler = new Handler(Looper.getMainLooper());

    @Inject
    public DevicesViewModel(DeviceRepository deviceRepository) {
        this.deviceRepository = deviceRepository;
        this.devices = deviceRepository.getDevices();
    }

    public LiveData<List<Device>> getDevices() { return devices; }
    public LiveData<DeviceActionState> getDeviceActionState() { return deviceActionState; }
    public LiveData<ProvisioningState> getProvisioningState() { return provisioningState; }
    public LiveData<Map<String, OtaState>> getOtaState() { return otaState; }

    public void registerDevice(String deviceId, String deviceName) {
        if (deviceId == null || deviceId.trim().isEmpty() || deviceName == null || deviceName.trim().isEmpty()) {
            deviceActionState.setValue(DeviceActionState.error("Device ID and Name cannot be empty"));
            return;
        }

        deviceActionState.setValue(DeviceActionState.loading());
        deviceRepository.registerDevice(deviceId, deviceName, new RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                deviceActionState.postValue(DeviceActionState.success("Device registered successfully"));
            }

            @Override
            public void onFailure(Exception e) {
                deviceActionState.postValue(DeviceActionState.error(e.getMessage() != null ? e.getMessage() : "Failed to register device"));
            }
        });
    }

    public void unregisterDevice(String deviceId) {
        deviceActionState.setValue(DeviceActionState.loading());
        deviceRepository.unregisterDevice(deviceId, new RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                deviceActionState.postValue(DeviceActionState.success("Device unregistered successfully"));
            }

            @Override
            public void onFailure(Exception e) {
                deviceActionState.postValue(DeviceActionState.error(e.getMessage() != null ? e.getMessage() : "Failed to unregister device"));
            }
        });
    }

    public void updateDeviceName(String deviceId, String name) {
        deviceRepository.updateDeviceName(deviceId, name, new RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {}
            @Override
            public void onFailure(Exception e) {}
        });
    }

    public void configureDeviceWifi(String deviceId, String ssid, String pass) {
        provisioningState.setValue(ProvisioningState.searching());
        handler.postDelayed(() -> {
            provisioningState.setValue(ProvisioningState.connecting());
            handler.postDelayed(() -> {
                provisioningState.setValue(ProvisioningState.success());
                handler.postDelayed(() -> provisioningState.setValue(ProvisioningState.idle()), 1500);
            }, 2500);
        }, 2000);
    }

    public void checkAndTriggerOtaUpdate(String deviceId) {
        setOtaStateForDevice(deviceId, OtaState.checking());

        handler.postDelayed(() -> {
            setOtaStateForDevice(deviceId, OtaState.downloading(0.1f));
            simulateOtaDownload(deviceId, 1);
        }, 2000);
    }

    private void simulateOtaDownload(String deviceId, final int step) {
        if (step <= 10) {
            handler.postDelayed(() -> {
                setOtaStateForDevice(deviceId, OtaState.downloading(step / 10f));
                simulateOtaDownload(deviceId, step + 1);
            }, 400);
        } else {
            setOtaStateForDevice(deviceId, OtaState.installing());
            handler.postDelayed(() -> {
                setOtaStateForDevice(deviceId, OtaState.success());
                handler.postDelayed(() -> {
                    Map<String, OtaState> currentMap = new HashMap<>(otaState.getValue() != null ? otaState.getValue() : new HashMap<>());
                    currentMap.remove(deviceId);
                    otaState.setValue(currentMap);
                }, 1500);
            }, 1500);
        }
    }

    private void setOtaStateForDevice(String deviceId, OtaState state) {
        Map<String, OtaState> currentMap = new HashMap<>(otaState.getValue() != null ? otaState.getValue() : new HashMap<>());
        currentMap.put(deviceId, state);
        otaState.setValue(currentMap);
    }

    public void clearActionState() {
        deviceActionState.setValue(DeviceActionState.idle());
    }

    // Custom State Classes
    public static class DeviceActionState {
        public enum Status { IDLE, LOADING, SUCCESS, ERROR }
        private final Status status;
        private final String message;

        private DeviceActionState(Status status, String message) {
            this.status = status;
            this.message = message;
        }

        public static DeviceActionState idle() { return new DeviceActionState(Status.IDLE, ""); }
        public static DeviceActionState loading() { return new DeviceActionState(Status.LOADING, ""); }
        public static DeviceActionState success(String msg) { return new DeviceActionState(Status.SUCCESS, msg); }
        public static DeviceActionState error(String err) { return new DeviceActionState(Status.ERROR, err); }

        public Status getStatus() { return status; }
        public String getMessage() { return message; }
    }

    public static class ProvisioningState {
        public enum Status { IDLE, SEARCHING, CONNECTING, SUCCESS, ERROR }
        private final Status status;
        private final String message;

        private ProvisioningState(Status status, String message) {
            this.status = status;
            this.message = message;
        }

        public static ProvisioningState idle() { return new ProvisioningState(Status.IDLE, ""); }
        public static ProvisioningState searching() { return new ProvisioningState(Status.SEARCHING, ""); }
        public static ProvisioningState connecting() { return new ProvisioningState(Status.CONNECTING, ""); }
        public static ProvisioningState success() { return new ProvisioningState(Status.SUCCESS, ""); }
        public static ProvisioningState error(String err) { return new ProvisioningState(Status.ERROR, err); }

        public Status getStatus() { return status; }
        public String getMessage() { return message; }
    }

    public static class OtaState {
        public enum Status { CHECKING, DOWNLOADING, INSTALLING, SUCCESS, ERROR }
        private final Status status;
        private final float progress;
        private final String error;

        private OtaState(Status status, float progress, String error) {
            this.status = status;
            this.progress = progress;
            this.error = error;
        }

        public static OtaState checking() { return new OtaState(Status.CHECKING, 0f, ""); }
        public static OtaState downloading(float progress) { return new OtaState(Status.DOWNLOADING, progress, ""); }
        public static OtaState installing() { return new OtaState(Status.INSTALLING, 0f, ""); }
        public static OtaState success() { return new OtaState(Status.SUCCESS, 0f, ""); }
        public static OtaState error(String err) { return new OtaState(Status.ERROR, 0f, err); }

        public Status getStatus() { return status; }
        public float getProgress() { return progress; }
        public String getError() { return error; }
    }
}

File: Aqua Guard/frontend/app/src/main/java/com/aquaguard/presentation/devices/DevicesViewModel.kt
package com.aquaguard.presentation.devices

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aquaguard.domain.model.Device
import com.aquaguard.domain.repository.DeviceRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class DevicesViewModel @Inject constructor(
    private val deviceRepository: DeviceRepository
) : ViewModel() {

    val devices: StateFlow<List<Device>> = deviceRepository.getDevices()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _deviceActionState = MutableStateFlow<DeviceActionState>(DeviceActionState.Idle)
    val deviceActionState = _deviceActionState.asStateFlow()

    fun registerDevice(deviceId: String, deviceName: String) {
        if (deviceId.isBlank() || deviceName.isBlank()) {
            _deviceActionState.value = DeviceActionState.Error("Device ID and Name cannot be empty")
            return
        }
        viewModelScope.launch {
            _deviceActionState.value = DeviceActionState.Loading
            deviceRepository.registerDevice(deviceId, deviceName)
                .onSuccess {
                    _deviceActionState.value = DeviceActionState.Success("Device registered successfully")
                }
                .onFailure {
                    _deviceActionState.value = DeviceActionState.Error(it.message ?: "Failed to register device")
                }
        }
    }

    fun unregisterDevice(deviceId: String) {
        viewModelScope.launch {
            _deviceActionState.value = DeviceActionState.Loading
            deviceRepository.unregisterDevice(deviceId)
                .onSuccess {
                    _deviceActionState.value = DeviceActionState.Success("Device unregistered successfully")
                }
                .onFailure {
                    _deviceActionState.value = DeviceActionState.Error(it.message ?: "Failed to unregister device")
                }
        }
    }

    fun updateDeviceName(deviceId: String, name: String) {
        viewModelScope.launch {
            deviceRepository.updateDeviceName(deviceId, name)
        }
    }

    // Wi-Fi provisioning state
    private val _provisioningState = MutableStateFlow<ProvisioningState>(ProvisioningState.Idle)
    val provisioningState = _provisioningState.asStateFlow()

    fun configureDeviceWifi(deviceId: String, ssid: String, pass: String) {
        viewModelScope.launch {
            _provisioningState.value = ProvisioningState.Searching
            kotlinx.coroutines.delay(2000)
            _provisioningState.value = ProvisioningState.Connecting
            kotlinx.coroutines.delay(2500)
            _provisioningState.value = ProvisioningState.Success
            kotlinx.coroutines.delay(1500)
            _provisioningState.value = ProvisioningState.Idle
        }
    }

    // OTA status state
    private val _otaState = MutableStateFlow<Map<String, OtaState>>(emptyMap())
    val otaState = _otaState.asStateFlow()

    fun checkAndTriggerOtaUpdate(deviceId: String) {
        viewModelScope.launch {
            val currentMap = _otaState.value.toMutableMap()
            currentMap[deviceId] = OtaState.Checking
            _otaState.value = currentMap

            kotlinx.coroutines.delay(2000)
            currentMap[deviceId] = OtaState.Downloading(0.1f)
            _otaState.value = currentMap

            for (progress in 1..10) {
                kotlinx.coroutines.delay(400)
                currentMap[deviceId] = OtaState.Downloading(progress / 10f)
                _otaState.value = currentMap
            }

            currentMap[deviceId] = OtaState.Installing
            _otaState.value = currentMap
            kotlinx.coroutines.delay(1500)

            currentMap[deviceId] = OtaState.Success
            _otaState.value = currentMap
            kotlinx.coroutines.delay(1500)
            currentMap.remove(deviceId)
            _otaState.value = currentMap
        }
    }

    fun clearActionState() {
        _deviceActionState.value = DeviceActionState.Idle
    }
}

sealed interface DeviceActionState {
    object Idle : DeviceActionState
    object Loading : DeviceActionState
    data class Success(val message: String) : DeviceActionState
    data class Error(val error: String) : DeviceActionState
}

sealed interface ProvisioningState {
    object Idle : ProvisioningState
    object Searching : ProvisioningState
    object Connecting : ProvisioningState
    object Success : ProvisioningState
    data class Error(val message: String) : ProvisioningState
}

sealed interface OtaState {
    object Checking : OtaState
    data class Downloading(val progress: Float) : OtaState
    object Installing : OtaState
    object Success : OtaState
    data class Error(val error: String) : OtaState
}
